src/cz/crcs/ectester/standalone/libs/jni/nettle.c

```diff
@@ -0,0 +1,512 @@
+    (*env)->ReleaseByteArrayElements(env, result, result_data, 0);
+    ecc_scalar_clear(&privScalar);
+    ecc_point_clear(&eccPubPoint);
+    ecc_point_clear(&resultPoint);
+    mpz_clear(x);
+    return result;
+}
+
+JNIEXPORT jobject JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeKeyAgreementSpi_00024Nettle_generateSecret___3B_3BLjava_security_spec_ECParameterSpec_2Ljava_lang_String_2(JNIEnv *env, jobject self, jbyteArray pubkey, jbyteArray privkey, jobject params, jstring algorithm) {
+    throw_new(env, "java/lang/UnsupportedOperationException", "Not supported.");
+    return NULL;
+}
+
+int signature_to_der(struct dsa_signature* signature, unsigned char *result, int byte_size) {
+    size_t rSize;
+    size_t sSize;
+    int wholeSize;
+
+    mpz_export(NULL, &rSize, 1, sizeof(unsigned char), 0, 0, signature->r);
+    mpz_export(NULL, &sSize, 1, sizeof(unsigned char), 0, 0, signature->s);
+    wholeSize = 2 + 2 + byte_size + 2 + byte_size;
+    if (wholeSize > 127) {
+        wholeSize -=1;
+        byte_size -=1;
+    }
+    if (!result) {
+        return wholeSize;
+    }
+
+	int diff = 0;
+	if (wholeSize < 128) {
+        result[0] = 0x30;
+        result[1] = wholeSize - 2;
+        result[2] = 0x02;
+        result[3] = byte_size;
+        mpz_export(NULL, &rSize, 1, sizeof(unsigned char), 0, 0, signature->r);
+        diff = byte_size - rSize;
+        for (int i = 0; i < diff; i++) {
+            result[4 + i] = 0x00;
+        }
+        mpz_export(result + 4 + diff, &rSize, 1, sizeof(unsigned char), 0, 0, signature->r);
+
+        result[4 + byte_size] = 0x02;
+        result[4 + byte_size + 1] = byte_size;
+        mpz_export(NULL, &sSize, 1, sizeof(unsigned char), 0, 0, signature->s);
+        diff = byte_size - sSize;
+        for (int i = 0; i < diff; i++) {
+            result[4 + i + byte_size + 2 + i] = 0x00;
+        }
+        mpz_export(result + 4 + byte_size + 2 + diff, &sSize, 1, sizeof(unsigned char), 0, 0, signature->s);
+        return wholeSize;
+    }
+    result[0] = 0x30;
+    result[1] = 129;
+    result[2] = wholeSize - 3;
+    result[3] = 0x02;
+    result[4] = byte_size;
+    mpz_export(NULL, &rSize, 1, sizeof(unsigned char), 0, 0, signature->r);
+    diff = byte_size - rSize;
+    for (int i = 0; i < diff; i++) {
+        result[5 + i] = 0x00;
+    }
+    mpz_export(result + 5 + diff, &rSize, 1, sizeof(unsigned char), 0, 0, signature->r);
+
+    result[5 + byte_size] = 0x02;
+    result[5 + byte_size + 1] = byte_size;
+
+
+    mpz_export(NULL, &sSize, 1, sizeof(unsigned char), 0, 0, signature->s);
+    diff = byte_size - sSize;
+    for (int i = 0; i < diff; i++) {
+       result[5 + i + byte_size + 2 + i] = 0x00;
+    }
+    mpz_export(result + 5 + byte_size + 2 + diff, &sSize, 1, sizeof(unsigned char), 0, 0, signature->s);
+    return wholeSize;
+
+}
+
+int der_to_signature(struct dsa_signature* signature, unsigned char* der) {
+    if (der[0] != 0x30) {
```

Comment:
This is not completely correct DER parsing and might cause issues, you could look at https://github.com/crocs-muni/ECTester/blob/master/src/cz/crcs/ectester/standalone/libs/jni/c_utils.c#L178 for an idea of how to parse the DER signature correctly.

Reply:
Solved in 6134523